services/datamanager/datasync/sync.go

```diff
@@ -120,50 +112,115 @@ func (s *syncer) Close() {
 	}
 }
 
-func (s *syncer) upload(ctx context.Context, path string) {
-	if s.progressTracker.inProgress(path) {
-		return
+// TODO: expose errors somehow
+// Sync uploads everything in queue until it is closed and emptied.
+func (s *syncer) SyncCaptureQueues(queues []*datacapture.Queue) {
+	for _, q := range queues {
+		s.syncQueue(s.cancelCtx, q)
 	}
+}
 
-	s.progressTracker.mark(path)
+func (s *syncer) syncQueue(ctx context.Context, q *datacapture.Queue) {
 	s.backgroundWorkers.Add(1)
 	goutils.PanicCapturingGo(func() {
+		ticker := time.NewTicker(s.syncInterval)
+		defer ticker.Stop()
 		defer s.backgroundWorkers.Done()
-		//nolint:gosec
-		f, err := os.Open(path)
-		if err != nil {
-			s.logger.Errorw("error opening file", "error", err)
-			return
-		}
-		defer func(f *os.File) {
-			err := f.Close()
-			if err != nil {
-				s.logger.Errorw("error closing file", "error", err)
+		for {
+			if err := ctx.Err(); err != nil {
+				if !errors.Is(err, context.Canceled) {
+					s.logger.Error(err)
+				}
+				return
 			}
-		}(f)
-
-		uploadErr := exponentialRetry(
-			ctx,
-			func(ctx context.Context) error { return s.uploadFile(ctx, s.client, f, s.partID) },
-			s.logger,
-		)
-		if uploadErr != nil {
-			s.logger.Error(uploadErr)
-			return
-		}
 
-		// Delete the file and indicate that the upload is done.
-		if err := os.Remove(path); err != nil {
-			s.logger.Errorw("error while deleting file", "error", err)
-		} else {
-			s.progressTracker.unmark(path)
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
```

Comment:
Every sync interval, we sync all captured data. Note that the queue is still being written to as this happens, so it will basically sync until it's "caught up", then wait until sync interval has passed to start draining the queue again.

Reply:
^ That suggestion wouldn't actually work. Would instead need a separate variable that tracks where to stop and can be moved forward whenever sync is retriggered.

Reply:
So my reasoning here was:

This is the case only when we are capturing data faster than we're syncing it (even while syncing - so data capture rate > bandwidth). If we're capturing data faster than we're syncing it, the quantity of local-and-yet-to-be-synced captured data will just constantly increase. No matter what, we'd basically have this ever increasing gap between <last captured> and <last synced> that would only stop when:

- Sync catches up (either because it speeds up, or capture is stopped/slowed)
- We run out of disk

In this case, stopping syncing to enforce ONLY_SYNC_DATA_BETWEEN_TICKER_N_AND_N+1 seems like it would only prevent syncing from catching up, and increase the chance that the user just runs out of disk, which doesn't seem like a desired/improved outcome

Reply:
It sort of seems like we'd just be choosing to rate limit ourselves in such a way that actually harms the higher level product goal (sync the data we've captured). Because if we're in this case the two options are either:
- Let sync catch up, so should let it run
- Crash or stop capturing

Reply:
Re-checking then about the `if next == nil` check below — could that just return rather than continuing to loop? Since `q.Pop()` locks and sets the next file.

Reply:
Since the queue only closes when the collector is closed, it seems the below would continuously wait for more files even after catching up.

Reply:
> Re-checking then about the if next == nil check below — could that just return rather than continuing to loop? Since q.Pop() locks and sets the next file.

If the queue isn't closed, it means the collector isn't closed, so it can still be writing data (even if none is ready yet). The close is essentialy how the writer to the queue tells the reader that no more data is coming, similar to how channels are closed.

This is a bug though - the nil case should break (returning it to the ticker, waiting for the next sync interval to pass), not continue. Fixed.

Reply:
Ah yep, meant break rather than return. That fixes the behavior of continuously syncing.

Agree that in the edge case that we're capturing faster than we're syncing, then the correct approach is to keep pulling from the queue as long as it's nonempty and syncing.

Reply:
Reading back through the tests to see if we already test this explicitly, but would want a test that checks that we're not syncing in the interim between tick1 and tick2 even as multiple capture files are being generated.